File: lib/features/listening/domain/entities/listening_passage.dart
import '../../../dictionary/domain/entities/app_context.dart';
import '../../../dictionary/domain/entities/language.dart';
import '../../../vocabulary/domain/entities/cefr_level.dart';

enum ListeningKind { conversation, talk }

final class ListeningTurn {
  const ListeningTurn({this.speaker, this.gender, required this.text});

  final String? speaker; // 'A' or 'B' for a conversation; null for a talk
  final String? gender; // 'male' or 'female'; null if the AI omitted it
  final String text;

  /// `speaker` / `text` match apps/web/src/lib/listeningPassage.ts's
  /// `ListeningTurn`. Web keeps per-speaker gender only in the passage-level
  /// `speakerGenders` map; Flutter also keeps it per turn, so `gender` is
  /// serialized here as an extra key (web ignores it) to round-trip cleanly.
  Map<String, dynamic> toJson() => {
        'speaker': speaker,
        'gender': gender,
        'text': text,
      };

  factory ListeningTurn.fromJson(Map<String, dynamic> json) => ListeningTurn(
        speaker: json['speaker'] as String?,
        gender: json['gender'] as String?,
        text: json['text'] as String? ?? '',
      );
}

final class ListeningQuestion {
  const ListeningQuestion({
    required this.question,
    required this.options,
    required this.correctIndex,
  });

  final String question;
  final List<String> options; // always 4 items
  final int correctIndex; // 0-3

  /// Keys match apps/web/src/lib/listeningPassage.ts's `ListeningQuestion`.
  Map<String, dynamic> toJson() => {
        'question': question,
        'options': options,
        'correctIndex': correctIndex,
      };

  factory ListeningQuestion.fromJson(Map<String, dynamic> json) =>
      ListeningQuestion(
        question: json['question'] as String? ?? '',
        options: List<String>.from(json['options'] as List? ?? const []),
        correctIndex: json['correctIndex'] as int? ?? 0,
      );
}

final class ListeningPassage {
  const ListeningPassage({
    required this.id,
    required this.kind,
    required this.turns,
    required this.questions,
    required this.level,
    required this.context,
    required this.targetLanguage,
    required this.generatedAt,
    this.speakerGenders = const {},
  });

  final String id;
  final ListeningKind kind;
  final List<ListeningTurn> turns;
  final List<ListeningQuestion> questions; // always 3 items
  final CEFRLevel level;
  final AppContext context;
  final Language targetLanguage;
  final DateTime generatedAt;

  /// Keyed by speaker key ('A', 'B', or 'solo' — see [speakerKey]), value is
  /// 'male' or 'female'. Derived once at parse time from each speaker's
  /// first-seen turn (see `ListeningPassageSource._parse`).
  final Map<String, String> speakerGenders;

  /// `kind` / `turns` / `questions` / `speakerGenders` match the persisted
  /// `ComprehensionItem` shape in apps/web/src/lib/savedListeningExercises.ts
  /// (`speakerGenders` is web's `Partial<Record<Speaker, SpeakerGender>>` —
  /// a `{speakerKey: "male"|"female"}` map). The web persists
  /// `level` / `context` / `targetLanguage` on the wrapping saved-exercise
  /// document, not on the item sub-object — serialized here too so the
  /// Flutter entity round-trips on its own.
  Map<String, dynamic> toJson() => {
        'id': id,
        'kind': kind.name,
        'turns': turns.map((t) => t.toJson()).toList(),
        'questions': questions.map((q) => q.toJson()).toList(),
        'speakerGenders': speakerGenders,
        'level': level.name,
        'context': context.name,
        'targetLanguage': targetLanguage.name,
        'generatedAt': generatedAt.toIso8601String(),
      };

  factory ListeningPassage.fromJson(Map<String, dynamic> json) =>
      ListeningPassage(
        id: json['id'] as String? ?? '',
        kind: json['kind'] != null
            ? ListeningKind.values.byName(json['kind'] as String)
            : ListeningKind.conversation,
        turns: (json['turns'] as List? ?? const [])
            .map((t) => ListeningTurn.fromJson(t as Map<String, dynamic>))
            .toList(),
        questions: (json['questions'] as List? ?? const [])
            .map((q) => ListeningQuestion.fromJson(q as Map<String, dynamic>))
            .toList(),
        speakerGenders: (json['speakerGenders'] as Map?)?.map(
              (k, v) => MapEntry(k as String, v as String),
            ) ??
            const {},
        level: json['level'] != null
            ? CEFRLevel.values.byName(json['level'] as String)
            : CEFRLevel.a1,
        context: json['context'] != null
            ? AppContext.values.byName(json['context'] as String)
            : AppContext.general,
        targetLanguage: json['targetLanguage'] != null
            ? Language.values.byName(json['targetLanguage'] as String)
            : Language.english,
        generatedAt: json['generatedAt'] != null
            ? DateTime.parse(json['generatedAt'] as String)
            : DateTime.fromMillisecondsSinceEpoch(0),
      );
}

/// Normalizes a raw turn speaker ('A', 'B', or null for a talk) into the key
/// [ListeningPassage.speakerGenders] and [assignVoices] use.
String speakerKey(String? speaker) => speaker ?? 'solo';

/// Deterministic, computed fresh each time it's needed (not cached on any
/// session state) — cheap and pure given the same passage. Walks distinct
/// speakers in order of first appearance; for each, takes the next unused
/// voice slot (1 or 2) of that speaker's declared gender. A speaker with no
/// declared gender (malformed AI response) defaults to 'female' — an
/// arbitrary but harmless choice, since the alternative (throwing) would
/// break an otherwise-usable passage over a cosmetic voice-picking detail.
/// Ports apps/web/src/lib/listeningPassage.ts's assignVoices() exactly.
Map<String, String> assignVoices(ListeningPassage passage) {
  final seen = <String>{};
  final order = <String>[];
  for (final t in passage.turns) {
    final key = speakerKey(t.speaker);
    if (seen.add(key)) order.add(key);
  }

  final nextSlotByGender = <String, int>{'male': 1, 'female': 1};
  final result = <String, String>{};
  for (final speaker in order) {
    final gender = passage.speakerGenders[speaker] ?? 'female';
    final slot = nextSlotByGender[gender]!;
    result[speaker] = '$gender$slot';
    nextSlotByGender[gender] =
        slot == 1 ? 2 : 1; // wraps back to 1 past 2 speakers of the same gender
  }
  return result;
}

File: apps/web/src/lib/listeningPassage.ts
import { LANGUAGE_LABELS, type TargetLanguage } from "./languages";
import { APP_CONTEXT_LABELS, type AppContext } from "./appContext";
import type { VocabRecord } from "./vocabRecords";

export type ListeningKind = "conversation" | "talk";
export type SpeakerGender = "male" | "female";
export type VoiceId = "male1" | "male2" | "female1" | "female2";
export type Speaker = "A" | "B" | "solo";
type CefrLevel = VocabRecord["cefrLevel"];

export interface ListeningTurn {
  speaker: "A" | "B" | null; // null for a talk's single speaker
  text: string;
}

export interface ListeningQuestion {
  question: string;
  options: string[]; // always 4
  correctIndex: number; // 0-3
}

export interface ListeningPassage {
  kind: ListeningKind;
  turns: ListeningTurn[];
  questions: ListeningQuestion[]; // always 3
  speakerGenders: Partial<Record<Speaker, SpeakerGender>>;
  level: CefrLevel;
  context: AppContext;
  targetLanguage: TargetLanguage;
}

const CEFR_LABELS: Record<CefrLevel, string> = {
  a1: "A1",
  a2: "A2",
  b1: "B1",
  b2: "B2",
  c1: "C1",
  c2: "C2",
};

// Ports lib/features/listening/data/sources/listening_passage_source.dart's
// _buildPrompt word-for-word, with one addition: each turn also carries a
// "gender" field (male/female), consistent per speaker letter across the
// whole passage — needed to pick an appropriate voice on the web (Flutter
// differentiates speakers via on-device TTS pitch instead, which has no
// equivalent on this app's self-hosted Piper backend).
export function buildListeningPassagePrompt(
  level: CefrLevel,
  context: AppContext,
  targetLanguage: TargetLanguage
): string {
  const languageLabel = LANGUAGE_LABELS[targetLanguage];
  const levelLabel = CEFR_LABELS[level];
  const contextLabel = APP_CONTEXT_LABELS[context];
  return (
    `You are creating a TOEIC-style listening exercise for a Vietnamese speaker ` +
    `learning ${languageLabel}, at ${levelLabel} level, in a ${contextLabel} ` +
    `register/setting. ` +
    `Randomly choose ONE of these two formats: ` +
    `(1) a CONVERSATION between exactly two speakers labeled "A" and "B" only ` +
    `(e.g. at an office, store, or while traveling), with 3 to 6 turns alternating ` +
    `between "A" and "B"; or ` +
    `(2) a TALK by a single speaker (e.g. an announcement, advertisement, or set of ` +
    `instructions), split into 2 to 4 turns, each with speaker set to null. ` +
    `For every turn, also declare "gender" as "male" or "female" for that turn's ` +
    `speaker — keep it consistent for the same speaker letter across the whole ` +
    `passage (speaker "A" is always the same gender in every one of its turns; ` +
    `same for "B"). A conversation may use two speakers of the same gender or two ` +
    `different genders — vary this across different generations. ` +
    `Then write exactly 3 multiple-choice questions in ${languageLabel} about ` +
    `the passage, each with exactly 4 answer options in ${languageLabel}, ` +
    `testing the main idea, a specific detail, or an implied meaning — never a ` +
    `fill-in-the-blank question. ` +
    `Respond with JSON only (no markdown, no code fences): ` +
    `{"kind": "conversation" or "talk", ` +
    `"turns": [{"speaker": "A" or "B" or null, "gender": "male" or "female", "text": "..."}], ` +
    `"questions": [{"question": "...", "options": ["...", "...", "...", "..."], ` +
    `"correctIndex": 0}]}`
  );
}

function speakerKey(speaker: "A" | "B" | null): Speaker {
  return speaker ?? "solo";
}

// Ports ListeningPassageSource._parse, plus derives speakerGenders from
// each speaker's first-seen turn — deliberately not re-derived per turn,
// so an AI response that's inconsistent about a speaker's gender on a
// later turn doesn't change which voice gets used mid-passage.
export function parseListeningPassage(
  json: Record<string, unknown>,
  level: CefrLevel,
  context: AppContext,
  targetLanguage: TargetLanguage
): ListeningPassage {
  const kind: ListeningKind = json.kind === "conversation" ? "conversation" : "talk";

  const rawTurns = Array.isArray(json.turns) ? json.turns : [];
  const turns: ListeningTurn[] = rawTurns.map((t) => {
    const tm = t as Record<string, unknown>;
    const speaker = tm.speaker === "A" || tm.speaker === "B" ? tm.speaker : null;
    return { speaker, text: typeof tm.text === "string" ? tm.text : "" };
  });

  const speakerGenders: Partial<Record<Speaker, SpeakerGender>> = {};
  for (const t of rawTurns) {
    const tm = t as Record<string, unknown>;
    const speaker = tm.speaker === "A" || tm.speaker === "B" ? tm.speaker : null;
    const key = speakerKey(speaker);
    if (key in speakerGenders) continue; // first-seen wins
    if (tm.gender === "male" || tm.gender === "female") {
      speakerGenders[key] = tm.gender;
    }
  }

  const rawQuestions = Array.isArray(json.questions) ? json.questions : [];
  const questions: ListeningQuestion[] = rawQuestions.map((q) => {
    const qm = q as Record<string, unknown>;
    return {
      question: typeof qm.question === "string" ? qm.question : "",
      options: Array.isArray(qm.options) ? qm.options.map(String) : [],
      correctIndex: typeof qm.correctIndex === "number" ? qm.correctIndex : 0,
    };
  });

  return { kind, turns, questions, speakerGenders, level, context, targetLanguage };
}

// Deterministic, computed once per session (not per synthesizeSpeech call).
// Walks distinct speakers in order of first appearance; for each, takes the
// next unused voice slot of that speaker's declared gender. A speaker with
// no declared gender (malformed AI response) defaults to "female" — an
// arbitrary but harmless choice, since the alternative (throwing) would
// break an otherwise-usable passage over a cosmetic voice-picking detail.
export function assignVoices(passage: ListeningPassage): Partial<Record<Speaker, VoiceId>> {
  const seen = new Set<Speaker>();
  const order: Speaker[] = [];
  for (const t of passage.turns) {
    const key = speakerKey(t.speaker);
    if (!seen.has(key)) {
      seen.add(key);
      order.push(key);
    }
  }

  const nextSlotByGender: Record<SpeakerGender, 1 | 2> = { male: 1, female: 1 };
  const result: Partial<Record<Speaker, VoiceId>> = {};
  for (const speaker of order) {
    const gender = passage.speakerGenders[speaker] ?? "female";
    const slot = nextSlotByGender[gender];
    result[speaker] = `${gender}${slot}` as VoiceId;
    nextSlotByGender[gender] = slot === 1 ? 2 : 1; // wraps back to 1 past 2 speakers of the same gender
  }
  return result;
}

// Ports ComprehensionSessionResult.correctCount, expressed as a ratio
// (correctCount / total) rather than a raw count, matching this app's
// other scoring functions (e.g. dictation.ts's computeDictationScore)
// returning a 0-1 fraction. A null (unanswered) entry never counts as
// correct even if correctIndex happens to be 0.
export function scoreComprehension(passage: ListeningPassage, selectedAnswers: (number | null)[]): number {
  if (passage.questions.length === 0) return 0;
  let correct = 0;
  for (let i = 0; i < passage.questions.length; i++) {
    if (selectedAnswers[i] !== null && selectedAnswers[i] === passage.questions[i].correctIndex) {
      correct++;
    }
  }
  return correct / passage.questions.length;
}
